File: src/test/kotlin/com/ciazhar/debouncer/SpamCheckerTest.java
package com.ciazhar.debouncer;

import com.ciazhar.debouncer.lib.svmchecker.model.LinearSVMClassifier;
import com.ciazhar.debouncer.lib.svmchecker.model.LinearSVMModel;
import com.hankcs.hanlp.classification.classifiers.IClassifier;

import java.io.*;

import static com.hankcs.hanlp.utility.Predefine.logger;

public class SpamCheckerTest {
    public static final String CORPUS_FOLDER = "data/dataset";
    /**
     * Model save path
     */
    public static final String MODEL_PATH = "data/svm-classification-model.ser";

    public static void main(String[] args) throws IOException
    {
        IClassifier classifier = new LinearSVMClassifier(trainOrLoadModel());
        predict(classifier, "This helps the smooth muscles in ");
        predict(classifier, "From andraallisonhuaa@email.com  Thu Sep 19 17:51:47 2002\n" +
                "Return-Path: <andraallisonhuaa@email.com>\n" +
                "Delivered-To: zzzz@localhost.spamassassin.taint.org\n" +
                "Received: from localhost (jalapeno [127.0.0.1])\n" +
                "\tby zzzzason.org (Postfix) with ESMTP id EE11116F03\n" +
                "\tfor <zzzz@localhost>; Thu, 19 Sep 2002 17:51:45 +0100 (IST)\n" +
                "Received: from jalapeno [127.0.0.1]\n" +
                "\tby localhost with IMAP (fetchmail-5.9.0)\n" +
                "\tfor zzzz@localhost (single-drop); Thu, 19 Sep 2002 17:51:45 +0100 (IST)\n" +
                "Received: from webnote.net (mail.webnote.net [193.120.211.219]) by\n" +
                "    dogma.slashnull.org (8.11.6/8.11.6) with ESMTP id g8JEuLC24080 for\n" +
                "    <zzzz@jmason.org>; Thu, 19 Sep 2002 15:56:31 +0100\n" +
                "Received: from dev-mail1.netsgo.com ([211.39.34.61]) by webnote.net\n" +
                "    (8.9.3/8.9.3) with ESMTP id PAA02126 for <zzzz@spamassassin.taint.org>;\n" +
                "    Thu, 19 Sep 2002 15:56:51 +0100\n" +
                "From: andraallisonhuaa@email.com\n" +
                "Received: from email-com.mr.outblaze.com (203.145.4.117) by\n" +
                "    dev-mail1.netsgo.com (6.5.007) id 3D6CAB5A0016F30F; Thu, 19 Sep 2002\n" +
                "    23:40:18 +0900\n" +
                "Message-Id: <000072146bff$00001431$000010db@email-com.mr.outblaze.com>\n" +
                "To: <Undisclosed.Recipients@webnote.net>\n" +
                "Subject: Hgh: safe and effective release of your own growth hormone!30543\n" +
                "Date: Thu, 19 Sep 2002 21:45:31 -0500\n" +
                "MIME-Version: 1.0\n" +
                "Reply-To: 100074c213aaa002@hotmail.com\n" +
                "1: X-Mailer: Microsoft Outlook Express 5.50.4522.1200\n" +
                "Content-Type: text/html; charset=\"iso-8859-1\"\n" +
                "Content-Transfer-Encoding: quoted-printable\n" +
                "\n" +
                "<html>\n" +
                "<head>\n" +
                "   <meta http-equiv=3D\"Content-Type\" content=3D\"text/html; charset=3Diso-8=\n" +
                "859-1\">\n" +
                "   <meta name=3D\"GENERATOR\" content=3D\"Mozilla/4.77 [en] (Windows NT 5.0; =\n" +
                "U) [Netscape]\">\n" +
                "   <meta name=3D\"ProgId\" content=3D\"FrontPage.Editor.Document\">\n" +
                "   <title>Lose weight while building lean muscle mass  and reversing the\n" +
                "ravages of aging all at once</title>\n" +
                "</head>\n" +
                "<body>\n" +
                " \n" +
                "<center><table BORDER WIDTH=3D\"52%\" >\n" +
                "<tr>\n" +
                "<td WIDTH=3D\"90%\" BGCOLOR=3D\"#CCCCFF\"><b><font face=3D\"Arial\"><font color=3D=\n" +
                "\"#FFFFFF\"><font size=3D+2></font></font></font></b> \n" +
                "<center><table BORDER=3D0 CELLSPACING=3D10 CELLPADDING=3D0 COLS=3D1 WIDTH=3D=\n" +
                "\"100%\" >\n" +
                "<tr>\n" +
                "<td WIDTH=3D\"100%\" BGCOLOR=3D\"#000000\">\n" +
                "<center><b><font face=3D\"Arial\"><font color=3D\"#FFFFFF\"><font size=3D+2>Hu=\n" +
                "man\n" +
                "Growth Hormone Therapy</font></font></font></b></center>\n" +
                "</td>\n" +
                "</tr>\n" +
                "</table></center>\n" +
                "\n" +
                "<center>\n" +
                "<p><b><font face=3D\"Arial\"><font color=3D\"#FFFFFF\"><font size=3D+1>Lose we=\n" +
                "ight\n" +
                "while building lean muscle mass</font></font></font></b>\n" +
                "<br><b><font face=3D\"Arial\"><font color=3D\"#FFFFFF\"><font size=3D+1> and\n" +
                "reversing the ravages of aging all at once.</font></font></font></b>\n" +
                "<p><font face=3D\"Arial\"><font color=3D\"#FFFFFF\"><font size=3D+0>Remarkable=\n" +
                " discoveries\n" +
                "about Human Growth Hormones (<b>HGH</b>) </font></font></font>\n" +
                "<br><font face=3D\"Arial\"><font color=3D\"#FFFFFF\"><font size=3D+0>are chang=\n" +
                "ing\n" +
                "the way we think about aging and weight loss.</font></font></font></center=\n" +
                ">\n" +
                "\n" +
                "<br> \n" +
                "<center><table WIDTH=3D\"442\" BGCOLOR=3D\"#FFFFFF\" >\n" +
                "<tr>\n" +
                "<td WIDTH=3D\"229\" HEIGHT=3D\"2\"><b><font face=3D\"Arial, Helvetica, sans-ser=\n" +
                "if\"><font color=3D\"#000000\"><font size=3D+0>Lose\n" +
                "Weight</font></font></font></b>\n" +
                "<br><b><font face=3D\"Arial, Helvetica, sans-serif\"><font color=3D\"#000000\"=\n" +
                "><font size=3D+0>Build\n" +
                "Muscle Tone</font></font></font></b>\n" +
                "<br><b><font face=3D\"Arial, Helvetica, sans-serif\"><font color=3D\"#000000\"=\n" +
                "><font size=3D+0>Reverse\n" +
                "Aging</font></font></font></b>\n" +
                "<br><b><font face=3D\"Arial, Helvetica, sans-serif\"><font color=3D\"#000000\"=\n" +
                "><font size=3D+0>Increased\n" +
                "Libido</font></font></font></b>\n" +
                "<br><b><font face=3D\"Arial, Helvetica, sans-serif\"><font color=3D\"#000000\"=\n" +
                "><font size=3D+0>Duration\n" +
                "Of Penile Erection</font></font></font></b></td>\n" +
                "\n" +
                "<td WIDTH=3D\"199\" HEIGHT=3D\"2\"><b><font face=3D\"Arial, Helvetica, sans-ser=\n" +
                "if\"><font color=3D\"#000000\"><font size=3D+0> New\n" +
                "Hair Growth</font></font></font></b>\n" +
                "<br><b><font face=3D\"Arial, Helvetica, sans-serif\"><font color=3D\"#000000\"=\n" +
                "><font size=3D+0> Improved\n" +
                "Memory</font></font></font></b>\n" +
                "<br><b><font face=3D\"Arial, Helvetica, sans-serif\"><font color=3D\"#000000\"=\n" +
                "> Improved\n" +
                "skin</font></font></b>\n" +
                "<br><b><font face=3D\"Arial, Helvetica, sans-serif\"><font color=3D\"#000000\"=\n" +
                "> New\n" +
                "Hair Growth</font></font></b>\n" +
                "<br><b><font face=3D\"Arial, Helvetica, sans-serif\"><font color=3D\"#000000\"=\n" +
                "> Wrinkle\n" +
                "Disappearance </font></font></b></td>\n" +
                "</tr>\n" +
                "</table></center>\n" +
                "\n" +
                "<center>\n" +
                "<p><b><font face=3D\"Arial\"><font color=3D\"#FFFFFF\"><a href=3D\"http://marke=\n" +
                "t.businessonlinenow.com/hgh/\">Visit\n" +
                "Our Web Site and Lean The Facts: Click Here</a></font></font></b></center>\n" +
                "\n" +
                "<p> </td>\n" +
                "</tr>\n" +
                "</table></center>\n" +
                "\n" +
                "</body>\n" +
                "</html>\n" +
                "\n" +
                "\n" +
                "\n" +
                "\n" +
                "\n");
    }

    private static void predict(IClassifier classifier, String text)
    {
        System.out.printf("《%s》 属于分类 【%s】\n", text, classifier.classify(text));
    }

    private static LinearSVMModel trainOrLoadModel() throws IOException
    {
        LinearSVMModel model = (LinearSVMModel) readObjectFrom(MODEL_PATH);
        if (model != null) return model;

        File corpusFolder = new File(CORPUS_FOLDER);
        if (!corpusFolder.exists() || !corpusFolder.isDirectory())
        {
            System.err.println("Tanpa corpus teks, harap baca format corpus dan unduhan corpus yang didefinisikan dalam IClassifier.train (java.lang.String)：" +
                    "https://github.com/hankcs/HanLP/wiki/%E6%96%87%E6%9C%AC%E5%88%86%E7%B1%BB%E4%B8%8E%E6%83%85%E6%84%9F%E5%88%86%E6%9E%90");
            System.exit(1);
        }

        IClassifier classifier = new LinearSVMClassifier();  // Buat classifier. Untuk fungsi lebih lanjut, silakan merujuk ke definisi antarmuka IClassifier.
        classifier.train(CORPUS_FOLDER);                     // Model yang terlatih mendukung ketekunan dan tidak harus dilatih lain kali.
        model = (LinearSVMModel) classifier.getModel();
        saveObjectTo(model, MODEL_PATH);
        return model;
    }

    /**
     * 序列化对象
     *
     * @param o
     * @param path
     * @return
     */
    public static boolean saveObjectTo(Object o, String path)
    {
        try
        {
            ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(path));
            oos.writeObject(o);
            oos.close();
        }
        catch (IOException e)
        {
            logger.warning("Pengecualian terjadi saat menyimpan objek "+ o +" ke "+ path +"" + e);
            return false;
        }

        return true;
    }

    /**
     * 反序列化对象
     *
     * @param path
     * @return
     */
    public static Object readObjectFrom(String path)
    {
        ObjectInputStream ois = null;
        try
        {
            ois = new ObjectInputStream(new FileInputStream(path));
            Object o = ois.readObject();
            ois.close();
            return o;
        }
        catch (Exception e)
        {
            logger.warning("Dari" + path + "Pengecualian terjadi saat membaca objek" + e);
        }

        return null;
    }
}

File: src/test/kotlin/com/ciazhar/debouncer/SpamCheckerTest.kt
package com.ciazhar.debouncer

import com.ciazhar.debouncer.lib.svmchecker.SpamChecker
import com.ciazhar.debouncer.lib.svmchecker.service.SVMCheckerService
import org.junit.Test
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.Files.isRegularFile
import jdk.nashorn.internal.objects.NativeArray.forEach
import java.util.function.Predicate
import java.util.stream.Stream
import io.vertx.ext.web.impl.Utils.readFileToString
import com.google.common.io.Files.isFile
import org.apache.commons.io.FileUtils
import java.io.File





class SpamCheckerTest {
    @Test
    fun spamCheck() {

        val text = "Return-Path: <tqznmzkkdnayh@getherbalnow.info>\n" +
                "Delivered-To: rait@bruce-guenter.dyndns.org\n" +
                "Received: (qmail 26496 invoked from network); 1 Feb 2005 00:25:05 -0000\n" +
                "Received: from localhost (localhost [127.0.0.1])\n" +
                "  by bruce-guenter.dyndns.org ([192.168.1.3]); 01 Feb 2005 00:25:05 -0000\n" +
                "Received: from zak.futurequest.net ([127.0.0.1])\n" +
                "  by localhost ([127.0.0.1])\n" +
                "  with SMTP via TCP; 01 Feb 2005 00:25:05 -0000\n" +
                "Received: (qmail 30245 invoked from network); 1 Feb 2005 00:25:04 -0000\n" +
                "Received: from dsl-200-95-90-28.prod-infinitum.com.mx (unknown [200.95.90.28])\n" +
                "  by zak.futurequest.net ([69.5.6.152])\n" +
                "  with SMTP via TCP; 01 Feb 2005 00:25:01 -0000\n" +
                "X-Message-Info: STRQ+pfol/442+u/C+48/15683577152822\n" +
                "Received: from smtp-brassiere.abalone.tqznmzkkdnayh@getherbalnow.info ([200.95.90.28]) by ko37-r05.tqznmzkkdnayh@getherbalnow.info with Microsoft SMTPSVC(5.0.8987.8828);\n" +
                "\t Mon, 31 Jan 2005 21:22:57 -0300\n" +
                "Received: from flop75.biometry.tqznmzkkdnayh@getherbalnow.info (dose75.tqznmzkkdnayh@getherbalnow.info [200.95.90.28])\n" +
                "\tby smtp-altair.utterance.tqznmzkkdnayh@getherbalnow.info (Postfix) with SMTP id 5CBU8A33VUEQ\n" +
                "\tfor <rait@bruce-guenter.dyndns.org>; Mon, 31 Jan 2005 20:23:57 -0400\n" +
                "X-Message-Info: FYZYN+%ND_LC_CHAR[1-3]65+nmt+R+7/4132393387359\n" +
                "Received: (qmail 35145 invoked by uid 138); Tue, 01 Feb 2005 01:18:57 +0100\n" +
                "Date: Mon, 31 Jan 2005 21:18:57 -0300\n" +
                "Message-Id: <35655717.33376@tqznmzkkdnayh@getherbalnow.info>\n" +
                "From: Jamar Toney <tqznmzkkdnayh@getherbalnow.info>\n" +
                "To: Rait <rait@bruce-guenter.dyndns.org>\n" +
                "Subject: Cheapest V.IA.G.R.A!! 70% Discount! \n" +
                "MIME-Version: 1.0 (produced by cretannagy 7.6)\n" +
                "Content-Type: multipart/alternative;\n" +
                "\tboundary=\"--5706099938399901743\"\n" +
                "Content-Length: 423\n" +
                "Lines: 27\n" +
                "\n" +
                "----5706099938399901743\n" +
                "Content-Type: text/plain;\n" +
                "\tcharset=\"iso-2816-8\"\n" +
                "Content-Transfer-Encoding: quoted-printable\n" +
                "Content-Description: what crisscross poetry\n" +
                "\n" +
                "Most places charge $18,\n" +
                "\n" +
                "we charge just $1\n" +
                "\n" +
                "You will never get cheaper v.i.a.g.r.a.! \n" +
                "\n" +
                "Order today before offer expires. \n" +
                "\n" +
                "Delivered world wide!\n" +
                "\n" +
                "http://dfbeky-wr.com/free/?sash99 \n" +
                "\n" +
                "\n" +
                "\n" +
                "\n" +
                "\n" +
                "optout link\n" +
                "http://dfbeky-wr.com/rm.php?sash99 \n" +
                "\n" +
                "\n" +
                "----5706099938399901743--\n"

        val classifier = SVMCheckerService(SpamChecker.trainOrLoadModel())
        SpamChecker.predict(classifier, text)
    }

    @Test
    fun confussionMatrix(){

        val classifier = SVMCheckerService(SpamChecker.trainOrLoadModel())

        var hamHam = 0
        var hamSpam = 0
        var spamHam = 0
        var spamSpam = 0

        val CORPUS_FOLDER = "data/test-dataset"
        val corpusFolder = File(CORPUS_FOLDER)
        if (!corpusFolder.exists() || !corpusFolder.isDirectory) {
            println("Tanpa corpus teks, harap baca format corpus dan unduhan corpus yang didefinisikan dalam IClassifier.train (java.lang.String)：" + "https://github.com/hankcs/HanLP/wiki/%E6%96%87%E6%9C%AC%E5%88%86%E7%B1%BB%E4%B8%8E%E6%83%85%E6%84%9F%E5%88%86%E6%9E%90")
            System.exit(1)
        }
        if (!corpusFolder.exists()) throw IllegalArgumentException(String.format("目录 %s 不存在", corpusFolder.getAbsolutePath()))
        if (!corpusFolder.isDirectory())
            throw IllegalArgumentException(String.format("目录 %s 不是一个目录", corpusFolder.getAbsolutePath()))
        val folders = corpusFolder.listFiles() ?: return

        folders.forEach {
            var category = it.name
            if (it.isFile()) return
            val files = it.listFiles() ?: return
            files.forEach {
                val content = FileUtils.readFileToString(it)

                if (category=="ham"){
                    if (SpamChecker.predict(classifier, content)=="ham"){
                        hamHam++
                    }else{
                        hamSpam++
                    }
                }else if (category=="spam"){
                    if (SpamChecker.predict(classifier, content)=="ham"){
                        spamHam++
                    }else{
                        spamSpam++
                    }
                }
            }
        }
        println("HAM HAM : $hamHam")
        println("HAM SPAM : $hamSpam")
        println("SPAM HAM : $spamHam")
        println("SPAM SPAM : $spamSpam")
    }
}